tpot/gp_deap.py

```diff
@@ -38,8 +38,37 @@
 import warnings
 import threading
 
-# Limit loops to generate a different individual by crossover/mutation
-MAX_MUT_LOOPS = 50
+def pick_two_individuals_eligible_for_crossover(population):
+    """ Pick two individuals from the population which can do crossover, that is, they share a primitive.
+
+    Parameters
+    ----------
+    population: array of individuals
+
+    Returns
+    ----------
+    tuple: (individual, individual)
+        Two individuals which are not the same, but share at least one primitive.
+    """
+    primitives_by_ind = [set([node.name for node in ind if isinstance(node, gp.Primitive)])
+                         for ind in population]
+    pop_as_str = [str(ind) for ind in population]
+    
+    eligible_pairs = [(i, i+1+j) for i, ind1_prims in enumerate(primitives_by_ind)
+                                 for j, ind2_prims in enumerate(primitives_by_ind[i+1:])
+                                 if not ind1_prims.isdisjoint(ind2_prims) and
+                                    pop_as_str[i] != pop_as_str[i+1+j]]
+
+    eligible_pairs += [(j, i) for (i,j) in eligible_pairs]
+    
+    if not eligible_pairs:
+        # what TODO? placeholder
+        return population[0], population[0]
```

Comment:
If there are no eligible pairs, then I think your current solution is fine. As you mentioned, it's not ideal because it's essentially a copy operation.

We could have it return `None, None` in this case and have the calling code know to interpret that as a "use mutation instead" signal. If we encapsulate the mutation code into a function, maybe that won't be so bad.